dbfopen.c

```diff
@@ -1436,17 +1436,17 @@ int SHPAPI_CALL DBFWriteAttributeDirectly(DBFHandle psDBF, int hEntity,
         /*      Assign all the record fields.                                   */
         /* -------------------------------------------------------------------- */
         int j;
-        if (STATIC_CAST(int, strlen(STATIC_CAST(char *, pValue))) >
+        if (STATIC_CAST(int, strlen(STATIC_CAST(const char *, pValue))) >
             psDBF->panFieldSize[iField])
             j = psDBF->panFieldSize[iField];
         else
         {
             memset(pabyRec + psDBF->panFieldOffset[iField], ' ',
                    psDBF->panFieldSize[iField]);
-            j = STATIC_CAST(int, strlen(STATIC_CAST(char *, pValue)));
+            j = STATIC_CAST(int, strlen(STATIC_CAST(const char *, pValue)));
         }
 
-        strncpy(
+        memcpy(
```

Comment:
Is this change relevant for C++ compilation?

Reply:
no, I don't think so. There was a (false positive) warning with one of the compilers used by GDAL about strncpy() possibly truncating when called from DBFWriteDateAttribute()

Reply:
https://github.com/OSGeo/gdal/actions/runs/8216223401/job/22470549905
```
In file included from /home/runner/work/gdal/gdal/ogr/ogrsf_frmts/shape/dbfopen_wrapper.cpp:5,
                 from /home/runner/work/gdal/gdal/build-alpine/ogr/ogrsf_frmts/shape/CMakeFiles/shapelib.dir/Unity/unity_0_cxx.cxx:13:
In function 'int gdal_DBFWriteAttributeDirectly(DBFHandle, int, int, const void*)',
    inlined from 'int DBFWriteDateAttribute(DBFHandle, int, int, const SHPDate*)' at /home/runner/work/gdal/gdal/ogr/ogrsf_frmts/shape/dbfopen.c:1548:38:
/home/runner/work/gdal/gdal/ogr/ogrsf_frmts/shape/dbfopen.c:1449:16: error: 'char* strncpy(char*, const char*, size_t)' output may be truncated copying between 0 and 8 bytes from a string of length 8 [-Werror=stringop-truncation]
 1449 |         strncpy(
      |         ~~~~~~~^
 1450 |             REINTERPRET_CAST(char *, pabyRec + psDBF->panFieldOffset[iField]),
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1451 |             STATIC_CAST(const char *, pValue), j);
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
```

Pull request in GDAL where I also test those changes: https://github.com/OSGeo/gdal/pull/9436

Reply:
Yes, I noticed this one as well today: https://github.com/OSGeo/shapelib/actions/runs/8215424311/job/22468877287

It happened when compiling as static libs.